### ISSUE TITLE
Warning about missing push notification entitlement

### ISSUE THREAD

**Issue by andersborgabiro** [phonegap/build#485]
Here's a longish thread about the fact that iTunes Connect thinks an app uses push notifications, but actually doesn't, unless something outside of the user code does somehow.

http://community.phonegap.com/nitobi/topics/pgb_3_5_missing_push_notification_entitlement

Apple accepts apps with this warning, but new publishers might be stumped by the warning issued, as this discussion indicates, and Apple might change this policy, stopping further publishing of PGB-generated apps, so please give priority to this now.

Regards,
Anders Borg

**Comment by lreading** [comment]
+1

**Comment by JamesBewley** [comment]
+1

**Comment by antogerva** [comment]
I can't even use TestFlight because of this issue and believe this can cause issue for InAppPurchase as well, if you aren't a publisher yet.

**Comment by ivanmobigis** [comment]
+1

**Comment by Wambosa** [comment]
+1

**Comment by Farufon** [comment]
Is there a scheduled solution date?  Next build?

**Comment by JamesBewley** [comment]
See the latest response in the community thread:

"This is now more than a warning from Apple. Now our app is stuck in REVIEW until this is fixed apparently. Up to this point, (for the last year while this issue persisted) Apple 'allowed' with a warning. Today, it seems not the case. At least in my case. 

Please provide a 'when' as this acknowledged issue has been going on for quite some time and from noting above, is affecting a lot of people. 

thank you"

I am shocked it has gotten to this stage, the community even provided the solution!

Is PhoneGap build still a viable option?

**Comment by Farufon** [comment]
+1

**Comment by bcolflesh** [comment]
+1

**Comment by jeremiahsa** [comment]
+1

**Comment by jamielob** [comment]
+1

**Comment by goya** [comment]
added the following preference:

<preference name="disable-push-notifications" value="1" /> 

(only for >= 3.9.2)

**this is just a stop gap for this issue**, ill post a blog on how we treat this going forward.

**Comment by jwander** [comment]
+1

**Comment by andersborgabiro** [comment]
I can't fathom why no one working with PGB responds. Is it that hard to fix once and for all?

**Comment by JamesBewley** [comment]
@andersborgabiro, I think @goya is your man at Adobe.

@goya, We don't care so much about a blog post, just desperately want to see this github ticket resolved in the shortest possible time to remove the risk of being unable to publish apps build with PGB.  What is the expected time for delivering a fix into your production environment?

**Comment by andersborgabiro** [comment]
Alrighty. That was not altogether clear.

@goya: The preference line in your post disappeared.

**Comment by goya** [comment]
added the following preference:

    <preference name="disable-push-notifications" value="1" /> 

(only for >= 3.9.2)

**this is just a stop gap for this issue**, ill post a blog on how we treat this going forward.

**Comment by Farufon** [comment]
So, currently this is not available for PGB online service?  It's still 3.9.1.

**Comment by Farufon** [comment]
@goya  Brent - are you saying this fix is IN the current PGB online service?  Or that you're going to write a blog to tell us when it's going to be in the service?

If not, when do you anticipate this fix (or better, version 3.9.2 or (cli-5.2.1) actually in place 'somewhere'?

kind regards
j.

**Comment by goya** [comment]
yep currently live.

**Comment by Farufon** [comment]
Oh, why I'm asking is that the version listed on PhoneGap Build shows:
PhoneGap (iOS / Android / Windows)
cli-5.2.0 (3.9.1 / 4.1.1 / 3.8.1)
...not 3.9.2 as you say we need greater than...
and most of all, we still get the notice from Apple.

**Comment by Farufon** [comment]
If it is live, can you point us where that would be?  It's not on the online build service that I can see.

**Comment by Farufon** [comment]
After using the tag you provided.  

Missing Push Notification Entitlement - Your app appears to include API used to register with the Apple Push Notification service, but the app signature's entitlements do not include the "aps-environment" entitlement. If your app uses the Apple Push Notification service, make sure your App ID is enabled for Push Notification in the Provisioning Portal, and resubmit after signing your app with a Distribution provisioning profile that includes the "aps-environment" entitlement. See "Provisioning and Development" in the Local and Push Notification Programming Guide for more information. If your app does not use the Apple Push Notification service, no action is required. You may remove the API from future submissions to stop this warning. If you use a third-party framework, you may need to contact the developer for information on removing the API.

**Comment by andersborgabiro** [comment]
When I republished an app yesterday without this tag I didn't get the warning. Maybe Apple only issues it when the app is published for the first time, or the condition is reversed?

**Comment by Farufon** [comment]
Did you use the ONLINE service to build?  https://build.phonegap.com/   The one that still says...
PhoneGap (iOS / Android / Windows)
cli-5.2.0 (3.9.1 / 4.1.1 / 3.8.1)
...not 3.9.2 as @goya said we need greater than...??

Just need to put this into context here.

**Comment by Farufon** [comment]
@goya said the change is online... (in response to my request if the ONLINE BUILD SERVICE was updated.)   Not local build.     He said it was active online.  BUT the online service that Adobe is pushing says 5.2.0 cli, and 3.9.1 for ios.   Is this not what he's referring to?

**Comment by andersborgabiro** [comment]
I used cli-5.2.0 (3.9.1 / 4.1.1 / 3.8.1).

And I used "phonegap remote build" that posts the code to PGB that then builds it.

@goya Do I need to step up cli-5.2.0?

**Comment by Farufon** [comment]
It's truly and amazing service.  No question.  And I'm shocked I can even get this far without xcode or even using the CLI version.  The apps we're creating are amazing.  If this 'notification' issue is the only issue we're having with Apple, I'm amazed.  Still, it sure would be great to nip this last little irritating thing as it would create a really clean submission.  Soooo close.

**Comment by andersborgabiro** [comment]
@Farufon I like it too, but I guess you at least have to use Application Loader.

**Comment by Farufon** [comment]
@andersborgabiro Does using the CLI version offer you more control?  Possibly being a developer, you feel more comfortable with further editing in xcode after you build to the platforms directory?

And would it be a better course, to future proof the app to use PGB to build towards an xcode version in case there are PGB issues in the future and you'd have an exploded view of your project in xcode?

Just curious.

**Comment by andersborgabiro** [comment]
The benefit of "phonegap remote build" is quicker code building (no zipping etc). I have a script per project that builds each one.

I only use Application Loader for iOS, not XCode.

I'm more of an Android developer, so at times I build locally for Android as well, e.g. to debug plugins. I try to avoid XCode.

**Comment by Farufon** [comment]
btw, did you get an answer from goya, re: 3.9.1 being active on the online build site, vs. the 3.9.2 he says we need to avoid the Apple Notifications?

I just push my new project to github, then pull project into phonegap build online, (the html version) and build there.

**Comment by andersborgabiro** [comment]
I don't know what @goya means either :). For it to work there needs to be a new cli version as well.

I don't use Github.

**Comment by klickagent** [comment]
I do have the same issue. I would really like to be able to set the DISABLE_PUSH_NOTIFICATIONS option inside the config.xml file. Thanks a lot for your help!

**Comment by Farufon** [comment]
@goya please respond!  Still very confused as to why you've said, 4 days ago that this (3.9.2) fix is online, but the build service still shows 3.9.1 (cli-5.2.0).   It appears you've abandoned this forum.

**Comment by Farufon** [comment]
Seems like Adobe totally abandoned this discussion after the hit-and-run comment about 3.9.2.

**Comment by Farufon** [comment]
@goya We've heard on the PhoneGap blog site:  The current is cli-5.4.1 This supports iOS9. 

Why are you unwilling to respond to everyone this post since your last statement about notifications we're trying to qualify??????????

**Comment by Farufon** [comment]
@andersborgabiro I'm new to this phonegap conversation on Github.  Is it standard practice that Adobe just ignores it, even after making statements?  Really hideous on their part to make statements that aren't true and then never follow up.   We were under the impression this was a site being contributed to by Adobe staff.

**Comment by wildabeast** [comment]
Hey guys cli-5.4.1 is live on build and uses cordova-ios 3.9.2. That version with this preference should disable push notifications:

    <preference name="disable-push-notifications" value="1" /> 

**Note**: cli-5.4.1 hasn't been officially announced on PGB because its still being tested, and could see fixes / changes. But you should be able to safely use the above solution for iOS. Additionally as mentioned above the `disable-push-notifications` preference might be changed in favour of another solution.

**Comment by Farufon** [comment]
Thank you @wildabeast.  It's painfully clear that the online build service is not worth the effort.  We'll only use the CLI now, unfortunately as the promise of build.phonegap.com is quite interesting.

**Comment by andersborgabiro** [comment]
PGB has worked fine over time, with some occasional major release hickups. This push notification situation has been a long-time issue though, for no obvious reason.

Adobe employees could be better at communicating with (and listening to) its customers though. The forum at build.phonegap.com is handled by volunteers, and I rarely see anyone from Adobe communicating via Github.

**Comment by Farufon** [comment]
Can someone explain this?  Is it possible to use the CLI - and avoid xcode?   

Meaning, can we use the remote (cloud) service, but still take advantage of the newer cli features and fixes rather than the older cli employed by PGB cloud service. 

phonegap remote build ios

doesn't this command negate the use of the new 5.4.1 cli, if the remote cloud service is running 5.2.0?

**Comment by andersborgabiro** [comment]
The cloud service is what it is. You can of course affect the versions of plugins being used, but not the core. Adobe handles updates of the core, including invoking the right phonegap.js or cordova.js when building.

I do use "phonegap remote build", so I can have my project on my PC and easily build it from there, but that's really just a command to ZIP the project and send it to be built at PGB.

I'm wondering, are you making too much of this? Apple accepts the push notification warning. My main concern has been that they might not in the future, but I published an app days ago with cli-5.2.0 and it went fine. Do you need anything else from 5.4.1 right now?

**Comment by Farufon** [comment]
Here's where I'm stuck.  When you say...

"I do use "phonegap remote build", so I can have my project on my PC and easily build it from there, but that's really just a command to ZIP the project and send it to be built at PGB."

Does that mean it's using the updated functions of the local 5.4.1 and just packaging it on PGB remotely. Or is it ignoring the updated functions (such as notifications issue) of the local 5.4.1 and using the remote 5.2.0 exclusively?

**Comment by andersborgabiro** [comment]
No it doesn't. It only ZIPs your project (html, js, css, png etc), not the PhoneGap/Cordova environment that you might have installed locally. "phonegap remote build" is just a simple wrapper for the PGB API.

But you can set cli-5.4.1 in config.xml now, as it's available at PGB. Whether it's to be 100% trusted I don't know, as it's not official yet.

But as I hinted, also 5.2.0 works fine, except for the push notification warning.

**Comment by Farufon** [comment]
eg: ?
    <preference name="phonegap-version" value="cli-5.4.1" />

**Comment by andersborgabiro** [comment]
Right.

**Comment by andersborgabiro** [comment]
To the PhoneGap Build guys: Is 5.4.1 something you stand behind as the official version?

**Comment by andersborgabiro** [comment]
I submitted an app yesterday with preference name="disable-push-notifications" value="1" and cli-5.4.1 and I still get the warning. I unpacked the ISA to see that the settings are firmly there.

**Comment by flannelhead** [comment]
@goya PGB just got updated to `cli-6.0.0`. What's the status of this issue in that version?

**Comment by mobileias** [comment]
we have also same issue.

Since last 8th Dec 2016 till now.

As previous just one day we have already created new build and it working fine.
Just after one day we rebuild with same type code and app give above type error.
Any one have solution please ?

We are using cli-6.3.0 phonegap version version and rebuild from 
https://build.phonegap.com/

**Comment by webenformasyon** [comment]
we have also same issue and push notification is critical for our customer

**Comment by odeditzhak** [comment]
Also seeing the same issue since last Dec 8th 2016.  Apps that had no problems before stopped working and cannot be submitted to iTunes Connect.

**Comment by philipmward** [comment]
I can confirm I'm also experiencing this warning confirmed since December 8th.

**Comment by goya** [comment]
what is an app id of an app someone can share that has this entitlement?  preferrably using an ios cordova version > 4.0.0

**Comment by goya** [comment]
... and doesnt have a plugin that adds the need for a push entitlement ;)

**Comment by odeditzhak** [comment]
Looks like anyone trying to build apps on PGB with push notifications has been experiencing this issue since Dec 8th. It basically means that currently PGB **does not** support push notifications. This is a very serious issue!
You can find some ugly workarounds involving hacking into the ipa files in this thread: https://forums.adobe.com/thread/2248424
Also found comments that version 1.9.x of the plugin solves the problem but is currently not supported in PGB.

**Comment by MulderWH** [comment]
We're experiencing the same problem here. Since this week we get no valid 'aps-environment' entitlement string found for application in the console using web inspector. This error appears when the device tries to register for push notifications. This issue only occurs on iOS10, because on iOS9 the device is registered properly. I think it's because of the recent changes for iOS10 and CocoaPods support: https://twitter.com/PhoneGapBuild/status/805162507803365376

**Comment by debugerr** [comment]
@goya Hi Brett. I'm seeing this issue in app 1265278. 

I've tried to upgrade to the notification plugin 1.9.x but that hits another PGB blocker with Cocoapods (https://github.com/phonegap/phonegap-plugin-push/issues/1463). However, I doubt this would actually solve the entitlements issue we are having.

**Comment by smogHead** [comment]
goya - What are you asking for. I have an app that needs the push entitlement and I am having to use iReSign to fix it....?

**Comment by alphajet** [comment]
+1

**Comment by Wouter33** [comment]
Guys, if you are really stuck, i whipped up a little service: https://appsigning.com. Got some questions earlier to resign people's apps, but it was taking up too much hours to keep doing it for free.

**Comment by odeditzhak** [comment]
Very impressive @Wouter33! Where there's a bug there's an opportunity ;) You should thank the PGB team for taking their time to fix critical bugs :))

**Comment by andersborgabiro** [comment]
@Wouter33 So you do it manually behind the scenes? Great job, but I still think Adobe should do something about the bug :).

**Comment by wildabeast** [comment]
Hey guys this should be fixed now... give it a shot.

**Comment by philipmward** [comment]
I was able to successfully build and deploy to TestFlight about an hour ago with no error message emails, and no issues with push notifications :)

**Comment by Wouter33** [comment]
Confirmed! Nice job!

**Comment by frederikbosch** [comment]
You could also use this for building with PGB. It works for us.

```xml
    <engine name="cordova" version=">=6.4.0"/>
    <engine name="cordova-ios" version="~4.3.0"/>

    <plugin name="phonegap-plugin-push" source="npm" spec="https://github.com/frederikbosch/phonegap-plugin-push#fix-aps-environment">
        <param name="SENDER_ID" value="99999"/> <!-- pgb -->
        <variable name="SENDER_ID" value="99999"/> <!-- cordova -->
    </plugin>
```

So you need cordova cli 6.4.0, cordova ios and 4.3.0 and the fix from PR #1469. And then you can get push working with Phonegap Build.
